cocotb/handle.py

```diff
@@ -530,11 +519,67 @@ def _range_iter(self, left, right):
 
     @NonHierarchyObject.value.getter
     def value(self):
-        # need to iterate over the sub-object
-        result = []
-        for x in range(len(self)):
-            result.append(self[x].value)
-        return result
+        """Reference to the value of the simulation object.
+
+        Getting and setting the current value of an array is done
+        by iterating through sub-handles in left-to-right order.
+
+        **Getting value**:
+
+        The returned list contains the values of each sub-handle
+        in the order indicated in the table:
+
+        +--------------+---------------------+--------------------------------------+
+        | Verilog      | VHDL                | Python list                          |
+        +==============+=====================+======================================+
+        | ``arr[4:7]`` | ``arr(4 to 7)``     | ``[arr[4], arr[5], arr[6], arr[7]]`` |
+        +--------------+---------------------+--------------------------------------+
+        | ``arr[7:4]`` | ``arr(7 downto 4)`` | ``[arr[7], arr[6], arr[5], arr[4]]`` |
+        +--------------+---------------------+--------------------------------------+
+
+        **Setting value**:
+
+        The indices in the list ``value`` will be assigned to sub-handles
+        in the order indicated in the table:
+
+        +--------------+---------------------+--------------------------------+
+        | Verilog      | VHDL                | Array values                   |
+        +==============+=====================+================================+
+        | ``arr[4:7]`` | ``arr(4 to 7)``     | - ``arr[4]`` gets ``value[0]`` |
+        |              |                     | - ...                          |
+        |              |                     | - ``arr[7]`` gets ``value[3]`` |
+        +--------------+---------------------+--------------------------------+
+        | ``arr[7:4]`` | ``arr(7 downto 4)`` | - ``arr[7]`` gets ``value[0]`` |
+        |              |                     | - ...                          |
+        |              |                     | - ``arr[7]`` gets ``value[3]`` |
+        +--------------+---------------------+--------------------------------+
+
+        .. warning::
+            Assigning a value to a sub-handle:
+
+            - **Wrong**: ``dut.some_array.value[0] = 1`` (Gets value as a list then updates index 0)
+            - **Right**: ``dut.some_array[0].value = 1``
+        """
+        # Don't use self.__iter__, because it has an unwanted `except IndexError`
+        return [
+            self[i]
+            for i in self._range_iter(self._range[0], self._range[1])
+        ]
+
+    @value.setter
+    def value(self, value):
+        """Assign value from a list of same length to an array in left-to-right order.
+        Index 0 of the list maps to the left-most index in the array.
+
+        See getter docstring above.
+        """
```

Comment:
This needs:
```suggestion
        """
        if type(self) is not NonHierarchyIndexableObject:
            # subclasses do not get this behavior
            return NonHierarchyObject.value.fset(value)
```

Reply:
Hmm. Actually, I think the fix is to implement `setimmediatevalue` instead of `value.setter`. That's automatically replaced by the derived classes.

Reply:
Changing the base class of `NonConstantObject` to `NonHierarchyObject` might help things here. I'm not sure of all of the implications though.

Reply:
`_range` is set to `None` if it's not an indexable object, which disables `__iter__` and causes `__getitem__` to throw an `IndexError`.

Reply:
My suggested change here is irrelevant - this doesn't actually matter because all it affects is the behavior of `NonConstantObject.value = ...`, and I already made a similar change to the getter in a previous commit.

No action is needed here. I think the implications are broad enough that I don't want to have to think about it as part of this PR!